storage/src/global_state/trie_store/cache/mod.rs

```diff
@@ -0,0 +1,374 @@
+                }
+                Self::Branch {
+                    pointer_block: new_pointer_block,
+                }
+            }
+            Trie::Extension { affix, pointer } => Self::Extension {
+                affix,
+                pointer: Box::new(CachePointer::Stored(pointer)),
+            },
+        }
+    }
+}
+
+// An in-memory cache for Trie nodes that is backed up by a store.
+pub struct TrieCache<'a, K, V, S> {
+    root: TrieCacheNode<K, V>,
+    store: &'a S,
+}
+
+impl<'a, K, V, S> TrieCache<'a, K, V, S>
+where
+    K: ToBytes + FromBytes + Clone + Eq,
+    V: ToBytes + FromBytes + Clone + Eq,
+    S: TrieStore<K, V> + 'a,
+{
+    pub fn new<T, E>(txn: &T, store: &'a S, root: &Digest) -> Result<Self, E>
+    where
+        T: Readable<Handle = S::Handle>,
+        S::Error: From<T::Error>,
+        E: From<S::Error> + From<bytesrepr::Error> + From<CacheError>,
+    {
+        match store.get(txn, root)? {
+            Some(node) => Ok(Self {
+                root: node.into(),
+                store,
+            }),
+            None => Err(CacheError::RootNotFound(*root).into()),
+        }
+    }
+
+    pub fn insert<T, E>(&mut self, key: K, value: V, txn: &T) -> Result<(), E>
+    where
+        T: Readable<Handle = S::Handle>,
+        S::Error: From<T::Error>,
+        E: From<S::Error> + From<bytesrepr::Error> + From<CacheError>,
+    {
+        let path: Vec<u8> = key.to_bytes()?;
+
+        let mut depth: usize = 0;
+        let mut current = &mut self.root;
+
+        while depth < path.len() {
+            match current {
+                TrieCacheNode::Branch { pointer_block } => {
+                    let index: usize = {
+                        assert!(depth < path.len(), "depth must be < {}", path.len());
+                        path[depth].into()
+                    };
+
+                    let pointer = &mut pointer_block[index];
+                    if let Some(next) = pointer {
+                        if depth == path.len() - 1 {
+                            let leaf = TrieCacheNode::Leaf { key, value };
+                            *next = CachePointer::InMem(leaf);
+                            return Ok(());
+                        } else {
+                            depth += 1;
+
+                            next.load_from_store::<_, _, E>(txn, self.store)?;
+                            if let CachePointer::InMem(next) = next {
+                                current = next;
+                            } else {
+                                unreachable!("Stored pointer should have been converted");
+                            }
+                        }
+                    } else {
+                        let leaf = TrieCacheNode::Leaf { key, value };
+                        let _ = std::mem::replace(pointer, Some(CachePointer::InMem(leaf)));
+                        return Ok(());
+                    }
```

Comment:
I remember having a discussion about this, but not the outcome - do we skip the extension if it would point to a leaf? I guess it makes sense as an optimization, but I'm not sure if this is what we're doing now, or if we intend to do it.